src/Language/PureScript/Scheme/CodeGen/Transpiler.hs

```diff
@@ -216,27 +224,36 @@ caseToScheme values caseAlternatives =
       replaceVariables (exprToScheme result)
                        (variablesToReplace valuesAndBinders)
 
-    -- Each identifier in a VarBinder will be replaced by its corresponding
-    -- value. E.g. `foo m n = m + n' will emit a function whose parameters are
-    -- v and v0, but the `m + n' expression will use `m' and `n' as variable.
-    -- We have to replace each occurrence of `m' with `v' and each occurrence
-    -- of `n' with `v0'.
+
+    variablesToReplace :: [(Expr Ann, Binder a)] -> [(Text, AST)]
     variablesToReplace valuesAndBinders =
-      concatMap go valuesAndBinders
+      concatMap (\(value, binder) -> go (exprToScheme value) binder)
+                valuesAndBinders
       where
-        go (_value, NullBinder _ann) = []
-        go (_value, LiteralBinder _ann _literal) = []
-        go (value, VarBinder _ann ident) = [(runIdent ident, exprToScheme value)]
+        go _value (NullBinder _ann) = []
+        go _value (LiteralBinder _ann _literal) = []
+
+        -- Each identifier in a VarBinder will be replaced by its corresponding
+        -- value. E.g. `foo m n = m + n' will emit a function whose parameters are
+        -- v and v0, but the `m + n' expression will use `m' and `n' as variable.
+        -- We have to replace each occurrence of `m' with `v' and each occurrence
+        -- of `n' with `v0'.
+        go value (VarBinder _ann ident) = [(runIdent ident, value)]
 
         -- For a ConstructorBinder we have to replace each identifier in its
         -- binders with an access to the right index of the vector in the cdr
         -- of the tagged pair holding the Constructor values.
-        go (value, ConstructorBinder _ann _typeName _constructorName binders) =
-          mapWithIndex (\i b -> (go' b, vectorRef (cdr (exprToScheme value))
-                                                  (IntegerLiteral i)))
+        go value (ConstructorBinder _ann _typeName _constructorName binders) =
+          mapWithIndex (\i b -> (go' b, vectorRef (cdr value) (IntegerLiteral i)))
                        binders
 
-        go _ = error "Not implemented"
+        -- We have to unfold the level of indirection introduced by the
```

Comment:
I am not sure what you mean by "level of indirection" here and above

Reply:
It is for the same reason we do it at line 171. Should I add a comment nonetheless?

Reply:
In our case the variable is `v`, but it gets referred as `bar` because of the NamedBinder. Maybe I can make the comment clearer?

Reply:
I think so, calling this "level of indirection" is a bit confusing without further context - it's important to clarify that it's about the variable names